### PR TITLE
Llama3.1 8b model training on Trn1 and EKS HyperPod

### PR DESCRIPTION
**Pull request by aruncs2005**
*Issue #, if available:*

*Description of changes:*
Added a new example for training llama 3.1 model on SageMaker HyperPod EKS and Trn1 instances. The example can be used to train either llama 3.1 8b or llama 3 8b models. 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by KeitaW** [comment]
Thank you for working on this! Could we move directory to `3.test_cases/neuronx-distributed/llama3/eks`?

File: 3.test_cases/21.Neuronx_llama3/llama3_train.yaml-template [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+              - name: NEURON_RT_ASYNC_EXEC_MAX_INFLIGHT_REQUESTS
+                value: "3"
+              - name: NEURON_FUSE_SOFTMAX
+                value: "1"
+              - name: NEURON_CC_FLAGS
+                value: "--model-type transformer --distribution-strategy=llm-training --cache_dir=${NEURON_CACHE_DIR}"
+              command: 
+                - torchrun
+                - --nproc_per_node=32
+                - --nnodes=$NUM_NODES
```
Comment (KeitaW) [inline 3.test_cases/21.Neuronx_llama3/llama3_train.yaml-template:149]: You may want to add `--pretrained_weight 1`
> Enable Pretrained Weights: By default, Neuron Distributed initiates training without using pretrained weights. To enable the use of pretrained weights, set the value of the  --pretrained_weight argument to 1.

Comment (KeitaW) [reply 3.test_cases/21.Neuronx_llama3/llama3_train.yaml-template:149]: https://github.com/aws-samples/awsome-distributed-training/tree/main/3.test_cases/neuronx-distributed/llama3/slurm#section-3-run-continual-pretraining-job-with-neuron-distributed

File: 3.test_cases/21.Neuronx_llama3/README.md [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+4. [ZeRO-1](https://awsdocs-neuron.readthedocs-hosted.com/en/latest/frameworks/torch/torch-neuronx/tutorials/training/zero1_gpt2.html#zero1-gpt2-pretraining-tutorial)
+
+
+## 0. Prerequisites
+
+### 0.1. EKS Cluster 
+Before running this training, you'll need to create an Amazon EKS or a SageMaker HyperPod EKS cluster with atleast 1 trn1.32xlarge/ trn1n.32xlarge. Instructions can be found in [1.architectures](../../1.architectures), the [aws-do-eks](https://bit.ly/do-eks) project, or the [eks-blueprints](https://github.com/aws-ia/terraform-aws-eks-blueprints) project.
+
+### 0.2 HF Access token
```
Comment (KeitaW) [inline 3.test_cases/21.Neuronx_llama3/README.md:19]: This part won't be necessary if we train from scratch?

File: 3.test_cases/21.Neuronx_llama3/README.md [hunk 3]
```diff
@@ -0,0 +1,97 @@
+## Train Llama 3.1 8B model using SageMaker HyperPod
```
Comment (KeitaW) [inline 3.test_cases/21.Neuronx_llama3/README.md:1]:
```suggestion
## Train Llama 3.1 8B model on Kubernetes
```
As the README mentioned, the guide can be used on both EKS and HyperPod EKS.
> Before running this training, you'll need to create an Amazon EKS or a SageMaker HyperPod EKS cluster with atleast 1 trn1.32xlarge/ trn1n.32xlarge.

**Comment by KeitaW** [comment]
I see the `config`/`src` of the original NxD repo are included in the PR. Those files are maintained by NxD team and frequently updated. Usually we cannot expect the codes works on certain version of NxD also works on another version of NxD.  When we introduce slurm test case, we had the discussion and ended up avoid having those dependencies in the first place. 
https://github.com/aws-samples/awsome-distributed-training/pull/383
We may want to do the same for Kuberenetes test case as well. For example, we may clone the NxD repository inside the container and grab all the related files under `https://github.com/aws-neuron/neuronx-distributed/tree/main/examples/training/llama`, instead of copying them from local directory.

File: 3.test_cases/21.Neuronx_llama3/README.md [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+### Pull the pytorch-training-neuronx image locally
+
+Login to ECR and pull the `pytorch-training-neuronx` image
+
+```sh
+region=us-east-2
+dlc_account_id=763104351884
+aws ecr get-login-password --region $region | docker login --username AWS --password-stdin $dlc_account_id.dkr.ecr.$region.amazonaws.com
+
+docker pull 763104351884.dkr.ecr.us-east-2.amazonaws.com/pytorch-training-neuronx:2.1.2-neuronx-py310-sdk2.19.1-ubuntu20.04
```
Comment (KeitaW) [inline 3.test_cases/21.Neuronx_llama3/README.md:39]:
```suggestion
docker pull ${dlc_account_id}.dkr.ecr.${region}.amazonaws.com/pytorch-training-neuronx:2.1.2-neuronx-py310-sdk2.19.1-ubuntu20.04
```

File: 3.test_cases/21.Neuronx_llama3/tokenize_data.yaml-template [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+            volumeMounts:
+              - name: shmem
+                mountPath: /dev/shm
+              - name: persistent-storage
+                mountPath: /fsx
+              - name: hyperpod
+                mountPath: /var/log/aws/clusters
+            resources:
+              requests:
+                aws.amazon.com/neuron: 16
```
Comment (KeitaW) [inline 3.test_cases/21.Neuronx_llama3/tokenize_data.yaml-template:41]: Perhaps you want to make it as variable too? Like the one for ` vpc.amazonaws.com/efa`.

File: 3.test_cases/21.Neuronx_llama3/llama3_train.yaml-template (hunk 1 above)
Comment (aruncs2005) [reply 3.test_cases/21.Neuronx_llama3/llama3_train.yaml-template:149]: We have issues with using pretrained weights right, so totally removed that parameter as its not as straight forward as setting that value. What do you think ?

**Comment by aruncs2005** [comment]
HI @KeitaW I have made the changes recommended can you review and let me know

**Comment by KeitaW** [comment]
Thanks @aruncs2005 , checking it out today.

**Review by KeitaW — APPROVED** [review]
The src files can be removed as well. Let me create another PR on top of it.

**Pull request merged**